### PR TITLE
Initial Usable Version

### PR DESCRIPTION
**Pull request by Abhinav-ark**

### CODE REVIEW

**Comment by Ashrockzzz2003** [comment]
Will review by tonight! Sorry.

File: cmd/configs.go [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+const course_url = base_url + "/xmlui/handle/123456789/"
+
+const code_btech = "150"
+const code_ba_communication = "893"
+const code_ma_communication = "894"
+const code_integrated_msc_ma = "903"
+const code_mca = "331"
+const code_msw = "393"
+const code_mtech = "279"
```
Comment (Ashrockzzz2003) [inline cmd/configs.go]: nit: Best practices. https://go.dev/doc/effective_go#constants


```suggestion
const (
	code_btech             = "150"
	code_ba_communication  = "893"
	code_ma_communication  = "894"
	code_integrated_msc_ma = "903"
	code_mca               = "331"
	code_msw               = "393"
	code_mtech             = "279"
)
```

File: cmd/helpers.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+import (
+	"fmt"
+	"log"
+    "github.com/anaskhan96/soup"
+	"os/exec"
+    "runtime"
+)
+
+// fetchHTML - fetches and parses HTML from the given URL
```
Comment (Ashrockzzz2003) [inline cmd/helpers.go:11]: Hiphens not needed. Just adds complexity. Better to use function name as first word in comment.
nit: End comments with `full-stop`.

```suggestion
// fetchHTML fetches and parses HTML from the given URL.
```

File: cmd/helpers.go [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    doc, err := soup.Get(url)
+    if err != nil {
+        fmt.Println("Error fetching the URL. Make sure you're connected to Amrita WiFi or VPN.")
+        return ""
+    }
+    return doc
+}
+
+
+// openBrowser - opens a URL in the default web browser
```
Comment (Ashrockzzz2003) [inline cmd/helpers.go:22]: Hiphens not needed. Just adds complexity. Better to use function name as first word in comment.
nit: End comments with full-stop.
```suggestion
// openBrowser opens a URL in the default web browser.
```

File: cmd/helpers.go [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+        err = exec.Command("rundll32", "url.dll,FileProtocolHandler", url).Start()
+    case "darwin":
+        err = exec.Command("open", url).Start()
+    }
+    if err != nil {
+        log.Fatal(err)
+    }
+}
+
+// Stack implementation using Go slices
```
Comment (Ashrockzzz2003) [inline cmd/helpers.go]: nit: End comments with `full-stop`.
```suggestion
// Stack implementation using Go slices.
```

File: cmd/helpers.go (hunk 2 above)
Comment (Ashrockzzz2003) [inline cmd/helpers.go]: nit: function name is enough. hiphens add confusion. end with `full-stop`.

```suggestion
// fetchHTML fetches and parses HTML from the given URL.
```

File: cmd/helpers.go (hunk 3 above)
Comment (Ashrockzzz2003) [inline cmd/helpers.go]: Same as above.

```suggestion
// openBrowser opens a URL in the default web browser.
```

File: cmd/helpers.go [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+type Stack struct {
+	items []string
+}
+
+// Push - adds an item to the stack
```
Comment (Ashrockzzz2003) [inline cmd/helpers.go]: Same as above.

```suggestion
// Push adds an item to the stack.
```

File: cmd/helpers.go [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+func (s *Stack) Push(item string) {
+	s.items = append(s.items, item)
+}
+
+// Pop - removes and returns the top item from the stack
```
Comment (Ashrockzzz2003) [inline cmd/helpers.go]: Same as above.

```suggestion
// Pop removes and returns the top item from the stack.
```

File: cmd/helpers.go [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+func (s *Stack) Pop() string {
+	if len(s.items) == 0 {
+		return ""
+	}
+	item := s.items[len(s.items)-1]
+	s.items = s.items[:len(s.items)-1]
+	return item
+}
+
+// IsEmpty - returns true if the stack is empty
```
Comment (Ashrockzzz2003) [inline cmd/helpers.go]: Same as above.

```suggestion
// IsEmpty returns true if the stack is empty.
```

File: cmd/helpers.go [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+func (s *Stack) IsEmpty() bool {
+	return len(s.items) == 0
+}
+
+// Peek - returns the top item without removing it
```
Comment (Ashrockzzz2003) [inline cmd/helpers.go]: Same as above.

```suggestion
// Peek returns the top item without removing it.
```

File: cmd/helpers.go [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+func (s *Stack) Peek() string {
+	if len(s.items) == 0 {
+		return ""
+	}
+	return s.items[len(s.items)-1]
+}
+
+// NewStack - creates a new stack
```
Comment (Ashrockzzz2003) [inline cmd/helpers.go]: Same as above.

```suggestion
// NewStack creates a new stack.
```

File: cmd/helpers.go [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+func NewStack() *Stack {
+	return &Stack{}
+}
+
+var stack = NewStack()
+
+
+
+
```
Comment (Ashrockzzz2003) [inline cmd/helpers.go]: Just one empty line at last is enough. Space optimization for code :)

File: cmd/helpers.go [hunk 11]
```diff
@@ -0,0 +1,80 @@
+package cmd
+
+import (
+	"fmt"
+	"log"
+    "github.com/anaskhan96/soup"
+	"os/exec"
+    "runtime"
+)
```
Comment (Ashrockzzz2003) [inline cmd/helpers.go:9]: Follow https://google.github.io/styleguide/go/best-practices.html#import-ordering for best-practices.

File: cmd/configs.go [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+
```
Comment (Ashrockzzz2003) [inline cmd/configs.go]: Also
Can we rename constants to be `ALL-CAPS`. Might improve readability.

File: cmd/root.go [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+func Execute() {
+    if err := rootCmd.Execute(); err != nil {
+        fmt.Println(err)
+        os.Exit(1)
+    }
+}
+
+// start function - equivalent to start() in Python
+func start() {
+    fmt.Println("Available Courses:\n")
```
Comment (Ashrockzzz2003) [inline cmd/root.go]: while `Println` already adds a `\n` can we use `backtik` for strings?

File: cmd/root.go [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+
+// Execute adds all child commands to the root command and sets flags appropriately.
+func Execute() {
+    if err := rootCmd.Execute(); err != nil {
+        fmt.Println(err)
+        os.Exit(1)
+    }
+}
+
+// start function - equivalent to start() in Python
```
Comment (Ashrockzzz2003) [inline cmd/root.go:28]:
```suggestion
// start lists the available courses and redirects to chosen course.
```

File: cmd/semTable.go [hunk 15]
```diff
@@ -0,0 +1,72 @@
+package cmd
+
+import (
+    "fmt"
+    "github.com/anaskhan96/soup"
+)
+
+// semTable function - equivalent to your Python code
```
Comment (Ashrockzzz2003) [inline cmd/semTable.go]: Is this comment relevant?

File: cmd/semTable.go [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+func semTable(url string) {
+    fmt.Println("Fetching semesters...")
+    resp := fetchHTML(url)  // Fetch the HTML content
+
+    // Check if the response is empty
+    if resp == "" {
```
Comment (Ashrockzzz2003) [inline cmd/semTable.go]: A more solid comparison?

```suggestion
    if len(resp) == 0 {
```

File: cmd/semTable.go [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+
+import (
+    "fmt"
+    "github.com/anaskhan96/soup"
+)
+
+// semTable function - equivalent to your Python code
+func semTable(url string) {
+    fmt.Println("Fetching semesters...")
+    resp := fetchHTML(url)  // Fetch the HTML content
```
Comment (Ashrockzzz2003) [inline cmd/semTable.go]: Comment not needed. The function name is self explanatory.
Also consider renaming `resp` to `res`. Familiar word.

File: cmd/semTable.go [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+
+    // Check if the response is empty
```
Comment (Ashrockzzz2003) [inline cmd/semTable.go]:
```suggestion
    // Check if the response `string` is empty.
```

File: cmd/semTable.go [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+    if resp == "" {
+        fmt.Println("Failed to fetch the HTML content. Exiting.")
+        return
+    }
+    
+    // Parse the HTML content using soup
```
Comment (Ashrockzzz2003) [inline cmd/semTable.go]:
```suggestion
    // Parse the HTML content using soup.
```

File: cmd/semTable.go [hunk 20]
```diff
@@ -0,0 +1,72 @@
...
+
+    ul := div.Find("ul")
+    li := ul.FindAll("li")
+
+    if len(li) == 0 {
+        fmt.Println("No semesters found on the page.")
+        return
+    }
+
+    // Display the found items
```
Comment (Ashrockzzz2003) [inline cmd/semTable.go]:
```suggestion
    // Display the found semesters.
```

File: cmd/semTable.go [hunk 21]
```diff
@@ -0,0 +1,72 @@
...
+        a := link.Find("a")
+        if a.Error == nil {
+            span := a.Find("span")
+            if span.Error == nil {
+                fmt.Printf("%d\t%s\n", i+1, span.Text())  // Extract the text from the span element
+            }
+        }
+    }
+
+    // Option to add "Back"
```
Comment (Ashrockzzz2003) [inline cmd/semTable.go]:
```suggestion
    // Option to add "Back".
```

File: cmd/semTable.go [hunk 22]
```diff
@@ -0,0 +1,72 @@
...
+    fmt.Printf("%d\tBack\n", len(li)+1)
+
+    // append to stack
```
Comment (Ashrockzzz2003) [inline cmd/semTable.go]: self-explantory. comment not needed I feel.

File: cmd/semTable.go [hunk 23]
```diff
@@ -0,0 +1,72 @@
...
+    // Display the found items
+    fmt.Printf("No\tSemesters\n")
+    for i, link := range li {
+        a := link.Find("a")
+        if a.Error == nil {
+            span := a.Find("span")
+            if span.Error == nil {
+                fmt.Printf("%d\t%s\n", i+1, span.Text())  // Extract the text from the span element
+            }
+        }
```
Comment (Ashrockzzz2003) [inline cmd/semTable.go]: Don't ignore the error!

```suggestion
        if a.Error != nil {
            fmt.Println(a.Error);
        }
        span := a.Find("span")
        if span.Error != nil {
            fmt.Println(a.Error)
        }

        // Extract the text from the span element.
        fmt.Printf("%d\t%s\n", i+1, span.Text())
```

**Review by Ashrockzzz2003 — CHANGES_REQUESTED** [review]
Address the changes.
Make similar changes to `year.go` too.

File: cmd/semTable.go (hunk 17 above)
Comment (Ashrockzzz2003) [reply cmd/semTable.go]: I meant for `fetchHTML`

File: cmd/logo.go [hunk 24]
```diff
@@ -0,0 +1,12 @@
+package cmd
+
+const logo_ascii string = `
```
Comment (Ashrockzzz2003) [inline cmd/logo.go]: Looks awesome @Abhinav-ark!
👏🏻

**Comment by Abhinav-ark** [comment]
Closes #2

**Review by Ashrockzzz2003 — APPROVED** [review]
LGTM now. Let's address remaining comments in next PR.

File: cmd/helpers.go (hunk 11 above)
Comment (FLASH2332) [reply cmd/helpers.go:9]: This code snippet seems to be updated in latest version but changes in import ordering has been made

File: cmd/root.go (hunk 13 above)
Comment (FLASH2332) [reply cmd/root.go]: While `fmt.Println` automatically adds a newline character (`\n`), using backticks in Go strings is useful for raw string literals. Raw string literals are enclosed in backticks (`` ` ``) and can include line breaks and other unescaped characters. 

For example:
```go
fmt.Println(`Available Courses:
- Course 1
- Course 2
- Course 3`)
```
In this case, backticks allow you to include the list directly within the string without needing escape characters.

For existing code with no special characters or line breaks, using double quotes is just as effective. But I have made the changes in the latest version as this has been mentioned here.

File: cmd/root.go (hunk 14 above)
Comment (FLASH2332) [reply cmd/root.go:28]: Comment has been updated.

File: cmd/semTable.go (hunk 16 above)
Comment (FLASH2332) [reply cmd/semTable.go]: This code snippet seems to be updated in latest version.

File: cmd/semTable.go (hunk 17 above)
Comment (FLASH2332) [reply cmd/semTable.go]: This code snippet has been updated.

File: cmd/semTable.go (hunk 20 above)
Comment (FLASH2332) [reply cmd/semTable.go]: Comment has been updated.

**Review by FLASH2332 — COMMENTED** [review]
The mentioned changes have been made. However, a few parts of the code seem to have moved to other files.

**Pull request merged**